sound/soc/sof/ipc.c

```diff
@@ -207,11 +207,23 @@ static int tx_wait_done(struct snd_sof_ipc *ipc, struct snd_sof_ipc_msg *msg,
 {
 	struct snd_sof_dev *sdev = ipc->sdev;
 	struct sof_ipc_cmd_hdr *hdr = msg->msg_data;
-	int ret;
+	int retries = 0;
+	int ret = 0;
 
 	/* wait for DSP IPC completion */
-	ret = wait_event_timeout(msg->waitq, msg->ipc_complete,
-				 msecs_to_jiffies(IPC_TIMEOUT_MS));
+	while (!ret && retries++ < 2) {
+		ret = wait_event_timeout(msg->waitq, msg->ipc_complete,
+					 msecs_to_jiffies(IPC_TIMEOUT_MS));
+		if (ret == 0) {
+			dev_warn(sdev->dev, "ipc timeout, polling irq thread\n");
+			if (sof_ops(sdev) && sof_ops(sdev)->irq_thread) {
+				disable_irq(sdev->ipc_irq);
+				ret = sof_ops(sdev)->irq_thread(sdev->ipc_irq,
+								sdev);
+				enable_irq(sdev->ipc_irq);
```

Comment:
It looks super strange to me for IPC framework directly call irq_thread function. For i.MX8 the IPC is wrapped around drivers/mailbox infrastructure and the client of the mailbox doesn't directly have access to IRQ handler. It only provides a callback IRQ handler will call. :)

Not sure if we can somehow abstract the IPC communication using the mailbox API.

Reply:
@dbaluta Thanks -- you are right this makes  assumptions about irq handling that are not part of the current ipc device abstractions. Fortunately it seems we can close this pull req anyways as another better solution was found. Probably if something similar is needed in the future, it needs to be done in the hw dependent backend, and generic ipc code jsut gets the snd_sof_ipc_reply().

@plbossart Ack, not needed anymore, I'll close this.